database/src/backend/heed/database.rs

```diff
@@ -200,59 +200,63 @@ impl<T: Table> DatabaseRw<T> for HeedTableRw<'_, '_, T> {
 
     #[inline]
     fn delete(&mut self, key: &T::Key) -> Result<(), RuntimeError> {
-        self.db.delete(&mut self.tx_rw.borrow_mut(), key)?;
-        Ok(())
+        match self.db.delete(&mut self.tx_rw.borrow_mut(), key) {
+            Ok(true) => Ok(()),
+            Ok(false) => Err(RuntimeError::KeyNotFound),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    #[inline]
+    fn take(&mut self, key: &T::Key) -> Result<T::Value, RuntimeError> {
+        // LMDB/heed does not return the value on deletion.
+        // So, fetch it first - then delete.
+        let value = get::<T>(&self.db, &self.tx_rw.borrow(), key)?;
+        match self.db.delete(&mut self.tx_rw.borrow_mut(), key) {
+            Ok(true) => Ok(value),
+            Err(e) => Err(e.into()),
+            // We just `get()`'ed the value - it is
+            // incorrect for it to suddenly not exist.
+            Ok(false) => unreachable!(),
+        }
     }
 
     #[inline]
     fn pop_first(&mut self) -> Result<(T::Key, T::Value), RuntimeError> {
         let tx_rw = &mut self.tx_rw.borrow_mut();
 
-        // Get the first value first...
-        let Some(first) = self.db.first(tx_rw)? else {
+        // Get the value first...
+        let Some((key, value)) = self.db.first(tx_rw)? else {
             return Err(RuntimeError::KeyNotFound);
         };
 
         // ...then remove it.
-        //
-        // We use an iterator because we want to semantically
-        // remove the _first_ and only the first `(key, value)`.
-        // `delete()` removes all keys including duplicates which
-        // is slightly different behavior.
-        let mut iter = self.db.iter_mut(tx_rw)?;
-
-        // SAFETY:
-        // It is undefined behavior to keep a reference of
-        // a value from this database while modifying it.
-        // We are deleting the value and never accessing
-        // the iterator again so this should be safe.
-        unsafe {
-            iter.del_current()?;
+        match self.db.delete(tx_rw, &key) {
+            Ok(true) => Ok((key, value)),
+            Err(e) => Err(e.into()),
+            // We just `get()`'ed the value - it is
+            // incorrect for it to suddenly not exist.
+            Ok(false) => unreachable!(),
```

Comment:
## `backend/heed`

Updated function bodies:
- `pop_first()`
- `pop_last()`

These now just use `heed::Database::first()/last()` + `delete()` instead of `unsafe` cursors.

The old comment doesn't matter since we don't care about duplicates anymore.